test/wallet/witness_address.cpp

```diff
@@ -0,0 +1,265 @@
+    const witness_address address(ec_public(test_public_key),
+        witness_address::testnet_bech32_p2wpkh,
+        witness_address::testnet_prefix);
+
+    BOOST_REQUIRE(address);
+    BOOST_REQUIRE_EQUAL(address.bech32(), TESTNET_P2WPKH_ADDRESS2);
+    BOOST_REQUIRE_EQUAL(address.encoded(), TESTNET_P2WPKH_ADDRESS2);
+}
+
+BOOST_AUTO_TEST_CASE(witness_address__witness_p2wpkh_vectors__valid_expected)
+{
+    for (const auto& test: native_witness_p2wpkh_tests)
+    {
+        short_hash hash;
+        BOOST_REQUIRE(decode_base16(hash, test.hash));
+        const witness_address address(hash, test.version);
+        BOOST_REQUIRE(address);
+        BOOST_REQUIRE_EQUAL(address.encoded(), test.address);
+    }
+}
+
+BOOST_AUTO_TEST_CASE(witness_address__witness_p2wsh_vectors__valid_expected)
+{
+    for (const auto& test: native_witness_p2wsh_tests)
+    {
+        hash_digest hash;
+        BOOST_REQUIRE(decode_base16(hash, test.hash));
+        const witness_address address(hash, test.version);
+        BOOST_REQUIRE(address);
+        BOOST_REQUIRE_EQUAL(address.encoded(), test.address);
+    }
+}
+
+struct witness_address_accessor
+  : public witness_address
+{
+    data_chunk convert_bits(uint32_t from_bits, uint32_t to_bits, bool pad,
+        const data_chunk& in, size_t in_offset)
+    {
+        return witness_address::convert_bits(from_bits, to_bits, pad, in,
+            in_offset);
+    }
+};
+
+BOOST_AUTO_TEST_CASE(base32__witness_address_zero_version_tests__valid_expected)
+{
+    for (const auto& test: witness_address_zero_version_tests)
+    {
+        base32 decoded;
+        BOOST_REQUIRE(decode_base32(decoded, test.address));
+        uint8_t witness_version = decoded.payload.front();
+        BOOST_REQUIRE(witness_version == 0);
+
+        witness_address_accessor converter;
+        const auto witness_program = converter.convert_bits(5, 8, false,
+            decoded.payload, 1);
+        BOOST_REQUIRE(witness_program.size() > 1 &&
+                      witness_program.size() < 41);
+        const auto witness_program_size = static_cast<uint8_t>(
+            witness_program.size());
+
+        const auto result = build_chunk(
+        {
+            to_chunk(witness_version),
+            to_chunk(witness_program_size),
+            witness_program
+        });
+
+        BOOST_REQUIRE_EQUAL(encode_base16(result), test.decoded);
+    }
+}
+
+BOOST_AUTO_TEST_CASE(base32__witness_address_nonzero_version_tests__valid_expected)
+{
+    // This op_version gap is used to normalize the value difference
+    // between the defined OP_0 opcode (value 0) and the OP_1 opcode
+    // (value 0x51)
+    const auto op_version_gap = 0x50;
+    const auto witness_v0_program_minimum_limit = 1;
+    const auto witness_v0_program_maximum_limit = 41;
+
+    for (const auto& test: witness_address_nonzero_version_tests)
+    {
+        base32 decoded;
+        BOOST_REQUIRE(decode_base32(decoded, test.address));
```

Comment:
I tend to group the statements such that `BOOST_REQUIRE(...)` (or a set of them) terminates the section, leaving a blank line after, then on to the next group.